### ISSUE TITLE
Clarify fd_seek behaviour for out-of-bounds access

### ISSUE THREAD

**Issue by RReverser** [WebAssembly/WASI#292]
`fd_seek` states that it's "similar to lseek".

However, the latter has some... interesting behaviour in regards to holes and random seek that 1) depends on the provided flags and 2) might be complicated to replicate across all platforms.

It would be helpful if the API docs would explicitly state whether out-of-bounds seeking is permitted and, if so, what is expected when it happens.

P.S. This is somewhat related to #9 which I hope will cover such edge-cases.

**Comment by sunfishcode** [comment]
Could you be more specific? POSIX says "The lseek() function shall allow the file offset to be set beyond the end of the existing data in the file. If data is later written at this point, subsequent reads of data in the gap shall return bytes with the value 0 until data is actually written into the gap", and at first glance, at least Linux and FreeBSD documentation look consistent with this.

**Comment by RReverser** [comment]
Yeah but then there are SEEK_DATA and SEEK_HOLE which fail on offset past the end of file; then there is also Windows SetFilePointer, which also allows seeking past the end of the file, but as per docs "leaves bytes uninitialised" rather than promises to zero-fill them.

In general, I just think it would be helpful to be explicit about what happens, as, at least as I'm reading it, "similar to" doesn't imply precisely same behaviour.

**Comment by sunfishcode** [comment]
WASI doesn't define `SEEK_DATA` or `SEEK_HOLE`, for that part. It's a good point that the obvious implementation of seeking past the end of a file on Windows leaves the bytes uninitialized. The question then is, should WASI require implementations on Windows to zero these bytes? It seems tempting, so that WASI can conform to POSIX in this respect, and to make WASI programs more portable in general, however we should evaluate the overhead of doing so.

"Similar to" here is meant to gloss over minor differences like `fd_seek` returning an error code instead of setting libc's `errno`. The intention is that we follow POSIX semantics except where explicitly documented otherwise, though as you've observed here, more work is needed to discover and document all the places where especially Windows makes POSIX semantics impractical.

It's also an interesting question whether we should seek to make WASI's documentation of POSIX-like APIs self-sufficient, or whether we should plan to always rely on POSIX as a normative reference (to which we can add exceptions and variations as needed). What do you think?

**Comment by RReverser** [comment]
Right, this is something I was thinking about several times when stumbling upon couple of methods with unclear semantics that referenced POSIX methods as "similar to".

While I understand that referencing POSIX is an easy choice compared to specifying all the possible edge cases, I think it's equally a question of the long-term WASI strategy & design - whether authors want it to be defined in terms of POSIX, or as an explicitly independent and universal API.

Right now it's mostly defined as the former option, but my personal view (which might not count much :) ) of WASI has always been the latter - that its goal is to be a cross-platform API that avoids platform-dependant semantics in the API, and, correspondingly, should define the expected behaviour without referring to potentially platform-specific docs.

You could argue that referring to POSIX is not "platform-specific" since it covers a family platforms, but it still assumes specific model of operating system and makes it hard to understand or implement WASI for non-POSIX-compatible systems.

**Summary:** I think there's a lot of value in defining semantics in WASI independently from any platform-specific docs and providing tests (#9) for the specified behaviour to make sure that it's indeed working in the same way on different platforms.

However, I realise that this requires more time and effort to document all the expectations, and in the short term it would be helpful to at least replace "similar to" references with explicit "behaves as" + flag mappings and expectations for edge cases.

**Comment by RReverser** [comment]
Just to add: `poll_oneoff` seems to have similar issue, except it doesn't even reference existing API for the reference behaviour.

**Comment by sunfishcode** [comment]
Yes; `poll_oneoff` has similarities to POSIX `poll`, though in that case the differences are greater. The `poll`'s `pollfd` struct differs from `poll_oneoff`'s `subscriptions` and `events`, and the way timeouts work is different too.

Documenting everything completely would indeed be great! It all comes down to finding volunteers to step up and do the work. If anyone's interested, I'd be happy to help provide pointers of where to start.

**Comment by RReverser** [comment]
> It all comes down to finding volunteers to step up and do the work.

Heh, to be fair I was rather hoping that, instead of external volunteers, participating implementors / companies would be able to dedicate time to this as part of the spec work :)
